mlflow/sklearn/__init__.py

```diff
@@ -455,6 +461,265 @@ def load_model(model_uri):
     )
 
 
+def _get_autolog_training_status():
+    return _autolog_training_status
+
+
+class _AutologTrainingStatus:
+    """
+    This class is designed for holding information which is used by autologging metrics
+    It will hold information of:
+    (1) a map of model_id to run_id, this information is registered during `patched_fit`
+    (2) a map of "prediction result object id" to a tuple of dataset name(the dataset is
+       the one which generate the prediction result) and model id.
+    (3) some scope flags which indicates currently in which method scope, we record scope
+       flags for method `model.fit`, `eval_and_log_metrics`, `model.score`.
+       In order to avoid nested/duplicated autologging metric, when run into these scopes,
+       we need temporarily disable the metric autologging.
+    (4) eval_dataset_info_map, it is a double level map:
+       `eval_dataset_info_map[model_uuid][eval_dataset_var_name]` will get a list, each
+       element in the list is a tuple of `(id(eval_dataset_instance), eval_dataset_sample_rows)`
+       This data structure is used for:
+        * generating unique dataset name key when autologging metric
+        * storing the dataset row samples, we need log it into metric_info artifact file.
+    (5) metric_api_call_arg_dict_list_map, it is a map, key is metric api function name,
+       and value is a list of call arguments dict.
+    """
+
+    fit_scope_key = 'fit'
+    eval_and_log_metrics_scope_key = 'eval_and_log_metrics'
+    model_score_scope_key = 'model_score'
```

Comment:
@WeichenXu123 can you take a look at https://github.com/mlflow/mlflow/pull/4491#discussion_r668429880 and let me know if I'm missing something? I saw that it was marked `Resolved` a couple days ago, but I didn't see a response to the comment & the change wasn't implemented. Reposting here:

"""
Can we use a single piece of state - `should_log_eval_metrics = True` - and have each environment disable eval metrics using this boolean? I don't think we need a separate piece of boolean state for each of these circumstances. Please let me know if I'm missing something.

Can we also create a context manager to use instead of try/finally? Something like

```
with disable_log_eval_metrics():
    # call model.score(), etc
```

I think it's okay to always collect dataset info when `predict()` is called. We can just use this boolean to control whether or not we log metrics within a particular context.
"""